hw/misc/hermes.c

```diff
@@ -57,11 +62,160 @@ const struct hermes_bar0 {
     .ehdssze = (1 * MiB)
 };
 
-typedef struct {
+typedef struct HermesState HermesState;
+
+/*
+ * BAR2 is used to do DMA using the xdma driver
+ * (https://github.com/aws/aws-fpga/tree/master/sdk/linux_kernel_drivers/xdma)
+ *
+ * The specification of BAR 2 can be found on the following document (Tables
+ * 35-126):
+ * https://www.xilinx.com/support/documentation/ip_documentation/xdma/v4_1/pg195-pcie-dma.pdf
+ * (note that in this document it uses BAR1, but here and on AWS we use BAR2)
+ */
+struct hermes_bar2_engine_reg {
+    uint32_t identifier;         /* 0x00 */
+    uint32_t control;            /* 0x04, 0x08 and 0x0C */
+    uint32_t status;             /* 0x40 and 0x44 */
+    uint32_t cmp_desc_count;     /* 0x48 */
+    uint32_t alignment;          /* 0x4C */
+    uint32_t wb_addr_low;        /* 0x88 */
+    uint32_t wb_addr_high;       /* 0x8C */
+    uint32_t inter_enable_mask;  /* 0x90, 0x94 and 0x98 */
+    uint32_t pmc;                /* 0xC0 */
+    uint32_t pcc0;               /* 0xC4 */
+    uint32_t pcc1;               /* 0xC4 */
+    uint32_t pdc0;               /* 0xCC */
+    uint32_t pdc1;               /* 0xD0 */
+};
+
+struct hermes_bar2_irq_reg {
+    uint32_t identifier;              /* 0x00 */
+    uint32_t user_inter_enable_mask;  /* 0x04, 0x08 and 0x0C */
+    uint32_t chnl_inter_enable_mask;  /* 0x10, 0x14 and 0x18 */
+    uint32_t user_inter_request;      /* 0x40 */
+    uint32_t chnl_inter_request;      /* 0x44 */
+    uint32_t user_inter_pending;      /* 0x48 */
+    uint32_t chnl_inter_pending;      /* 0x4C */
+    uint32_t user_vct_num0;           /* 0x80 */
+    uint32_t user_vct_num1;           /* 0x84 */
+    uint32_t user_vct_num2;           /* 0x88 */
+    uint32_t user_vct_num3;           /* 0x8C */
+    uint32_t chnl_vct_num0;           /* 0xA0 */
+    uint32_t chnl_vct_num1;           /* 0xA4 */
+};
+
+struct hermes_bar2_cfg_reg {
+    uint32_t identifier;          /* 0x00 */
+    uint32_t busdev;              /* 0x04 */
+    uint32_t pcie_mpl;            /* 0x08 */
+    uint32_t pcie_mrrs;           /* 0x0C */
+    uint32_t sysid;               /* 0x10 */
+    uint32_t msi_enable;          /* 0x14 */
+    uint32_t pcie_data_w;         /* 0x18 */
+    uint32_t pcie_ctrl;           /* 0x1C */
+    uint32_t axi_usr_mpl;         /* 0x40 */
+    uint32_t axi_usr_mrrs;        /* 0x44 */
+    uint32_t write_flush_timeout; /* 0x60 */
+};
+
+struct hermes_bar2_sgdma_reg {
+    uint32_t identifier;     /* 0x00 */
+    uint32_t desc_low_addr;  /* 0x80 */
+    uint32_t desc_high_addr; /* 0x84 */
```

Comment:
Given that these are internal to the model, I see no reason why you can't store the address as a uint64... or even hwaddr immediately instead of storing an upper and a lower and combining them at use.

Reply:
Makes sense, fixed